utils/s3/client.cc

```diff
@@ -1092,6 +1093,117 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::chunked_download_source final : public seastar::data_source_impl {
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+    range _range;
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    std::deque<temporary_buffer<char>> _buffers;
+    size_t _buffers_size = 0;
+    semaphore _buffers_sem;
+    bool _is_closing = false;
+    bool _is_done = false;
+    seastar::condition_variable _buffer_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        while (true) {
+            try {
+                s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+                if ((_as && _as->abort_requested()) || _is_closing) {
+                    s3l.trace("Abandoning fiber for object '{}'", _object_name);
+                    co_return;
+                }
+                auto download_source = _client->make_download_source(_object_name, _range, _as);
```

Comment:
You ask for a full range here, but once receive _max_buffers_size you just close the source. Why not ask for at most the size you're going to download?

Reply:
Coz I give up the fiber in each loop cycle, at this moment a buffer could be stealed by `get()` then it means on the next iteration I'm going to underperform since I set this artificial limit of the whole loop to halt at  `_max_buffers_size`. So, the optimistic approach is to set `len` to the real `len` and not to limit it by anything that is less than this.

Reply:
I don't get it. Consider the initial range is [0:1Gb). First loop here you call for [0:1Gb) range and start fetching buffers from it. Once you fetch 5Mb the download source is closed and another loop starts with [5Mb;1Gb) range. Once another 5Mb are fetched, source is closed, next is created with [10Mb;1Gb) range.

Or maybe I'm misreading the code :(

In case I'm not -- what's the point in requesting [0;1Gb) in the first attempt if you're not going to read it past 5Mb? And what's the point in requesting [5Mb;1Gb) in the second, if the bytes after 10Mb are not going to be .get()-ed? And so on.

Reply:
> Once you fetch 5Mb the download source is closed...

Or not...
In the loop `while (_buffers_size < _max_buffers_size && !_is_closing)` the `_buffers_size < _max_buffers_size` may always be true since it has `coroutine::maybe_yield()` in the end of every iteration which may transfer the execution to `get()` which will drain the queue so the `while` will keep reading indefinitely until all the data is drained